libsolidity/formal/CHC.cpp

```diff
@@ -1208,6 +1371,13 @@ void CHC::defineExternalFunctionInterface(FunctionDefinition const& _function, C
 	m_context.addAssertion(smt::symbolicUnknownConstraints(state().balance(state().thisAddress()) + k.currentValue(), TypeProvider::uint256()));
 	state().addBalance(state().thisAddress(), k.currentValue());
 
+	if (encodeExternalCallsAsTrusted())
+		// If the contract has state variables that are addresses to other contracts,
+		// we need to encode the fact that those contracts may have been called in between
+		// transactions to _contract.
+		// We do that by adding nondet_interface constraints for those contracts.
+		addNondetCalls(*m_currentContract);
```

Comment:
Generally, any contract may have been called in between transactions to ``_contract``...
But maybe only the ones in the state variables are relevant, is that the idea here? What if one of those contracts contains yet another contract in one of *its* state variables?

Would it be possible to encode *arbitrary* external calls to *any* contract for which ``isActive`` is true or is that impossible to encode? That's what you're approximating here, right?

Reply:
> What if one of those contracts contains yet another contract in one of its state variables?

It would be possible to also change those indirectly.
Say contract B has a state variable of type contract C, and A has one of B.
When we add these constraints (`addNondetCalls`) for a function from A, it adds a "nondeterministic call" to its B. B's nondeterministic interface allows calling any external function in B. Every external function in B will have the same call above (`addNondetCalls`), which would allow changes in C. Having said all that, I'm not sure it's enough. C could also change without B changing, but then it wouldn't matter for A. Maybe any changes on C that are relevant for A have to go through B.

In any case, maybe it's better to have a more general rule based on `isActive`, as you pointed out. I'll think about it. Nice catch!

Reply:
Yeah, you're right, thinking about it, I can't find an obvious case in which this wouldn't be enough, so maybe it's just fine!

Reply:
Yea, let me try a few examples first and add them to the PR.